File: python/ccxt/nobitex.py
from ccxt.base.exchange import Exchange
from ccxt.base.errors import AuthenticationError, OrderNotFound, NotSupported

class nobitex(Exchange):
    def describe(self):
        return self.deep_extend(super(nobitex, self).describe(), {
            'id': 'nobitex', 'name': 'Nobitex', 'countries': ['IR'], 'rateLimit': 100, 'certified': False,
            'has': {
                'CORS': None, 'spot': True, 'margin': False, 'swap': False, 'future': False,
                'fetchMarkets': True, 'fetchTicker': True, 'fetchTickers': True, 'fetchOrderBook': True,
                'fetchOHLCV': False, 'fetchTrades': True, 'fetchBalance': True, 'createOrder': True,
                'fetchOrder': True, 'cancelOrder': True, 'fetchOpenOrders': True, 'fetchClosedOrders': True,
                'fetchOrders': True, 'fetchMyTrades': True, 'fetchTransactions': True,
                'fetchTradingFee': False, 'fetchTradingFees': False,
            },
            'urls': {'logo': 'https://nobitex.ir/assets/images/logo.svg', 'api': {'public': 'https://apiv2.nobitex.ir', 'private': 'https://apiv2.nobitex.ir'}, 'www': 'https://nobitex.ir', 'doc': ['https://apidocs.nobitex.ir']},
            'api': {
                'public': {'get': ['market/stats', 'v2/trades/{symbol}', 'v3/orderbook/{symbol}', 'v2/orderbook/{symbol}', 'status']},
                'private': {
                    'get': ['users/profile', 'users/transactions-history', 'users/markets/favorite'],
                    'post': ['users/wallets/list', 'market/orders/add', 'market/orders/cancel', 'market/orders/list', 'market/orders/update-status', 'users/accounts-add']
                }
            },
            'requiredCredentials': {'apiKey': True, 'secret': False},
            'options': {'defaultType': 'spot'}
        })

    def sign(self, path, api='public', method='GET', params={}, headers=None, body=None):
        url = self.urls['api'][api] + '/' + self.implode_params(path, params)
        query = self.omit(params, self.extract_params(path))
        if headers is None: headers = {}
        headers['User-Agent'] = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        if api == 'private':
            if not self.apiKey: raise AuthenticationError('Nobitex requires apiKey')
            headers['Authorization'] = 'Token ' + self.apiKey
            if method == 'POST': headers['Content-Type'] = 'application/json'; body = self.json(query)
        elif method == 'GET':
            if query: url += '?' + self.urlencode(query)
        return {'url': url, 'method': method, 'body': body, 'headers': headers}

    def fetch_markets(self, params={}):
        response = self.request('market/stats', 'public', 'GET', params)
        stats = self.safe_value(response, 'stats', {})
        result = []
        for id, market_data in stats.items():
            parts = id.split('-')
            if len(parts) != 2: continue
            base, quote = parts[0], parts[1]
            result.append({
                'id': id, 'symbol': base + '/' + quote, 'base': base, 'quote': quote,
                'baseId': base, 'quoteId': quote, 'type': 'spot', 'spot': True, 'active': True,
                'precision': {'amount': 8, 'price': 1},
                'limits': {'amount': {'min': 0.00000001, 'max': None}, 'price': {'min': 0.00000001, 'max': None}, 'cost': {'min': None, 'max': None}},
                'info': market_data
            })
        return result

    def parse_ticker(self, ticker, market=None):
        ts = self.milliseconds()
        return self.safe_ticker({
            'symbol': self.safe_string(market, 'symbol'), 'timestamp': ts, 'datetime': self.iso8601(ts),
            'high': self.safe_number(ticker, 'dayHigh'), 'low': self.safe_number(ticker, 'dayLow'),
            'bid': self.safe_number(ticker, 'bestBuy'), 'ask': self.safe_number(ticker, 'bestSell'),
            'last': self.safe_number(ticker, 'latest'), 'close': self.safe_number(ticker, 'latest'),
            'baseVolume': self.safe_number(ticker, 'volumeSrc'), 'quoteVolume': self.safe_number(ticker, 'volumeDst'),
            'info': ticker
        }, market)

    def fetch_ticker(self, symbol, params={}):
        self.load_markets()
        market = self.market(symbol)
        response = self.request('market/stats', 'public', 'GET', self.extend({'srcCurrency': market['baseId'], 'dstCurrency': market['quoteId']}, params))
        return self.parse_ticker(self.safe_value(self.safe_value(response, 'stats', {}), market['baseId'] + '-' + market['quoteId'], {}), market)

    def fetch_tickers(self, symbols=None, params={}):
        self.load_markets()
        result = {}
        for s, m in self.markets.items():
            if self.safe_value(m, 'info') is not None: result[s] = self.parse_ticker(m['info'], m)
        return result

    def fetch_ohlcv(self, symbol, timeframe='1d', since=None, limit=None, params={}):
        raise NotSupported('Nobitex API does not support OHLCV data.')

    def fetch_balance(self, params={}):
        self.load_markets()
        response = self.request('users/wallets/list', 'private', 'POST', params)
        result = {'info': response}
        for w in self.safe_value(response, 'wallets', []):
            code = self.safe_currency_code(self.safe_string(w, 'currency'))
            total = self.safe_number(w, 'balance')
            free = self.safe_number(w, 'activeBalance')
            result[code] = {'free': free, 'used': (total - free if total is not None and free is not None else None), 'total': total}
        return self.safe_balance(result)

    def parse_order_status(self, status):
        return {'open':'open','active':'open','done':'closed','filled':'closed','canceled':'canceled','cancelled':'canceled'}.get(status, status)

    def parse_order(self, order, market=None):
        ts = self.safe_integer(order, 'createdAt')
        return self.safe_order({
            'id': self.safe_string(order, 'id'), 'clientOrderId': None, 'info': order, 'timestamp': ts, 'datetime': self.iso8601(ts),
            'status': self.parse_order_status(self.safe_string(order, 'status')), 'symbol': self.safe_string(market, 'symbol'),
            'type': self.safe_string(order, 'type'), 'side': self.safe_string_lower(order, 'side'), 'price': self.safe_number(order, 'price'),
            'amount': self.safe_number(order, 'amount'), 'filled': self.safe_number(order, 'filledAmount')
        }, market)

    def create_order(self, symbol, type, side, amount, price=None, params={}):
        self.load_markets()
        m = self.market(symbol)
        req = {'type': side, 'srcCurrency': m['baseId'], 'dstCurrency': m['quoteId'], 'amount': str(amount)}
        if type == 'limit' and price is not None: req['price'] = str(price)
        elif type == 'market': req['mode'] = 'market'
        resp = self.request('market/orders/add', 'private', 'POST', self.extend(req, params))
        if self.safe_string(resp, 'status') != 'ok': raise Exception(f"Nobitex Error: {self.safe_string(resp, 'code')} - {self.safe_string(resp, 'message')}")
        return self.parse_order(self.safe_value(resp, 'order', resp), m)

    def cancel_order(self, id, symbol=None, params={}):
        return self.request('market/orders/update-status', 'private', 'POST', self.extend({'order': id, 'status': 'canceled'}, params))

    def fetch_orders(self, symbol=None, since=None, limit=None, params={}):
        self.load_markets()
        m = self.market(symbol) if symbol else None
        req = {}
        if m: req.update({'srcCurrency': m['baseId'], 'dstCurrency': m['quoteId']})
        return self.parse_orders(self.safe_value(self.request('market/orders/list', 'private', 'POST', self.extend(req, params)), 'orders', []), m, since, limit)

    def fetch_open_orders(self, symbol=None, since=None, limit=None, params={}):
        return self.fetch_orders(symbol, since, limit, self.extend({'status': 'open'}, params))
    def fetch_closed_orders(self, symbol=None, since=None, limit=None, params={}):
        return self.fetch_orders(symbol, since, limit, self.extend({'status': 'done'}, params))
    def fetch_my_trades(self, symbol=None, since=None, limit=None, params={}):
        return self.fetch_orders(symbol, since, limit, self.extend({'status': 'done'}, params))

    def parse_transaction(self, transaction, currency=None):
        timestamp = self.parse8601(self.safe_string(transaction, 'created_at'))
        type = self.safe_string(transaction, 'type')
        amount = self.safe_number(transaction, 'amount')
        currency_code = self.safe_currency_code(self.safe_string(transaction, 'currency'))
        return {
            'info': transaction,
            'id': self.safe_string(transaction, 'id'),
            'txid': self.safe_string(transaction, 'txHash'),
            'timestamp': timestamp,
            'datetime': self.iso8601(timestamp),
            'type': type,
            'amount': amount,
            'currency': currency_code,
            'status': 'done',
            'fee': None,
            'address': None,
            'network': None,
        }

    
    def fetch_transactions(self, code=None, since=None, limit=None, params={}):
        self.load_markets()
        response = self.request('users/transactions-history', 'private', 'GET', params)
        return self.parse_transactions(self.safe_value(response, 'transactions', []), code, since, limit)

File: js/src/nobitex.js
import Exchange from './abstract/nobitex.js';
import { OrderNotFound, NotSupported } from './base/errors.js';
export default class nobitex extends Exchange {
    describe() {
        return this.deepExtend(super.describe(), {
            'id': 'nobitex',
            'name': 'Nobitex',
            'countries': ['IR'],
            'rateLimit': 100,
            'certified': false,
            'has': {
                'CORS': undefined,
                'spot': true,
                'margin': false,
                'swap': false,
                'future': false,
                'option': false,
                'fetchMarkets': true,
                'fetchTicker': true,
                'fetchTickers': true,
                'fetchOrderBook': true,
                'fetchOHLCV': false,
                'fetchTrades': true,
                'fetchBalance': true,
                'createOrder': true,
                'fetchOrder': true,
                'cancelOrder': true,
                'fetchOpenOrders': true,
                'fetchClosedOrders': true,
                'fetchOrders': true,
                'fetchMyTrades': true,
                'fetchTradingFee': false,
                'fetchTradingFees': false,
            },
            'urls': {
                'logo': 'https://nobitex.ir/assets/images/logo.svg',
                'api': {
                    'public': 'https://apiv2.nobitex.ir',
                    'private': 'https://apiv2.nobitex.ir',
                },
                'www': 'https://nobitex.ir',
                'doc': ['https://apidocs.nobitex.ir'],
            },
            'api': {
                'public': {
                    'get': [
                        'v2/trades/{symbol}',
                        'v3/orderbook/{symbol}',
                        'market/stats',
                        'status',
                    ],
                },
                'private': {
                    'get': [
                        'users/profile',
                        'users/transactions-history',
                        'users/markets/favorite',
                    ],
                    'post': [
                        'users/wallets/list',
                        'market/orders/add',
                        'market/orders/cancel',
                        'market/orders/list',
                        'market/orders/update-status',
                        'users/accounts-add',
                    ],
                },
            },
            'requiredCredentials': {
                'apiKey': true,
                'secret': false,
            },
            'options': {
                'defaultType': 'spot',
            },
        });
    }
    sign(path, api = 'public', method = 'GET', params = {}, headers = {}, body = undefined) {
        let url = this.urls['api'][api] + '/' + this.implodeParams(path, params);
        const query = this.omit(params, this.extractParams(path));
        if (Object.keys(query).length) {
            if (method === 'GET') {
                url += '?' + this.urlencode(query);
            }
        }
        if (Object.keys(headers).length === 0) {
            headers = {};
        }
        headers['User-Agent'] = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36';
        if (api === 'private') {
            this.checkRequiredCredentials();
            headers['Authorization'] = 'Token ' + this.apiKey;
            if (method === 'POST') {
                headers['Content-Type'] = 'application/x-www-form-urlencoded';
            }
        }
        return { 'url': url, 'method': method, 'body': body, 'headers': headers };
    }
    async fetchMarkets(params = {}) {
        const response = await this.request('market/stats', 'public', 'GET', params);
        const stats = this.safeDict(response, 'stats', {});
        const keys = Object.keys(stats);
        const markets = [];
        for (let i = 0; i < keys.length; i++) {
            const id = keys[i];
            const parts = id.split('-');
            if (parts.length !== 2) {
                continue;
            }
            const baseId = parts[0];
            const quoteId = parts[1];
            const base = this.safeCurrencyCode(baseId);
            const quote = this.safeCurrencyCode(quoteId);
            markets.push({
                'id': id,
                'symbol': base + '/' + quote,
                'base': base,
                'quote': quote,
                'settle': undefined,
                'baseId': baseId,
                'quoteId': quoteId,
                'settleId': undefined,
                'type': 'spot',
                'spot': true,
                'margin': false,
                'swap': false,
                'future': false,
                'option': false,
                'active': true,
                'contract': false,
                'linear': undefined,
                'inverse': undefined,
                'subType': undefined,
                'taker': undefined,
                'maker': undefined,
                'contractSize': undefined,
                'expiry': undefined,
                'expiryDatetime': undefined,
                'strike': undefined,
                'optionType': undefined,
                'precision': {
                    'amount': this.parseNumber('0.000001'),
                    'price': this.parseNumber('1'),
                    'cost': undefined,
                },
                'limits': {
                    'leverage': { 'min': undefined, 'max': undefined },
                    'amount': { 'min': undefined, 'max': undefined },
                    'price': { 'min': undefined, 'max': undefined },
                    'cost': { 'min': undefined, 'max': undefined },
                },
                'marginModes': {
                    'isolated': false,
                    'cross': false,
                },
                'created': undefined,
                'info': stats[id],
            });
        }
        return markets;
    }
    parseTicker(ticker, market = undefined) {
        const timestamp = this.milliseconds();
        const symbol = this.safeString(market, 'symbol');
        return this.safeTicker({
            'symbol': symbol,
            'timestamp': timestamp,
            'datetime': this.iso8601(timestamp),
            'high': this.safeNumber(ticker, 'dayHigh'),
            'low': this.safeNumber(ticker, 'dayLow'),
            'bid': this.safeNumber(ticker, 'bestBuy'),
            'bidVolume': undefined,
            'ask': this.safeNumber(ticker, 'bestSell'),
            'askVolume': undefined,
            'vwap': undefined,
            'open': this.safeNumber(ticker, 'dayOpen'),
            'close': this.safeNumber(ticker, 'latest'),
            'last': this.safeNumber(ticker, 'latest'),
            'previousClose': undefined,
            'change': undefined,
            'percentage': this.safeNumber(ticker, 'dayChange'),
            'average': undefined,
            'baseVolume': this.safeNumber(ticker, 'volumeSrc'),
            'quoteVolume': this.safeNumber(ticker, 'volumeDst'),
            'info': ticker,
        }, market);
    }
    async fetchTicker(symbol, params = {}) {
        await this.loadMarkets();
        const market = this.market(symbol);
        const request = {
            'srcCurrency': market['baseId'],
            'dstCurrency': market['quoteId'],
        };
        const response = await this.request('market/stats', 'public', 'GET', this.extend(request, params));
        const stats = this.safeDict(response, 'stats', {});
        const tickerKey = market['baseId'] + '-' + market['quoteId'];
        const ticker = this.safeDict(stats, tickerKey, {});
        return this.parseTicker(ticker, market);
    }
    async fetchTickers(symbols = undefined, params = {}) {
        await this.loadMarkets();
        const result = {};
        for (const symbol in this.markets) {
            const market = this.markets[symbol];
            const info = this.safeDict(market, 'info');
            if (info !== undefined) {
                result[symbol] = this.parseTicker(info, market);
            }
        }
        return result;
    }
    async fetchOrderBook(symbol, limit = undefined, params = {}) {
        await this.loadMarkets();
        const market = this.market(symbol);
        const requestSymbol = market['baseId'].toUpperCase() + market['quoteId'].toUpperCase();
        const response = await this.request('v3/orderbook/' + requestSymbol, 'public', 'GET', params);
        return this.parseOrderBook(response, market['symbol']);
    }
    async fetchOHLCV(symbol, timeframe = '1d', since = undefined, limit = undefined, params = {}) {
        throw new NotSupported(this.id + ' fetchOHLCV() is not supported by the Nobitex API.');
    }
    async fetchTrades(symbol, since = undefined, limit = undefined, params = {}) {
        await this.loadMarkets();
        const market = this.market(symbol);
        const requestSymbol = market['baseId'].toUpperCase() + market['quoteId'].toUpperCase();
        const response = await this.request('v2/trades/' + requestSymbol, 'public', 'GET', params);
        return this.parseTrades(response, market, since, limit);
    }
    async fetchBalance(params = {}) {
        await this.loadMarkets();
        const response = await this.request('users/wallets/list', 'private', 'POST', params);
        const wallets = this.safeList(response, 'wallets', []);
        const result = { 'info': response };
        for (let i = 0; i < wallets.length; i++) {
            const wallet = wallets[i];
            const currencyId = this.safeString(wallet, 'currency');
            const code = this.safeCurrencyCode(currencyId);
            // استفاده از safeNumber به جای safeString و تفریق استاندارد
            const total = this.safeNumber(wallet, 'balance');
            const free = this.safeNumber(wallet, 'activeBalance');
            let used = undefined;
            if (total !== undefined && free !== undefined) {
                used = total - free;
            }
            result[code] = {
                'free': free,
                'used': used,
                'total': total,
            };
        }
        return this.safeBalance(result);
    }
    parseOrderStatus(status) {
        const statuses = {
            'open': 'open',
            'active': 'open',
            'done': 'closed',
            'filled': 'closed',
            'canceled': 'canceled',
            'cancelled': 'canceled',
        };
        return this.safeString(statuses, status, status);
    }
    parseOrder(order, market = undefined) {
        const id = this.safeString(order, 'id');
        const timestamp = this.safeTimestamp(order, 'createdAt');
        const symbol = this.safeString(market, 'symbol');
        return this.safeOrder({
            'id': id,
            'clientOrderId': undefined,
            'info': order,
            'timestamp': timestamp,
            'datetime': this.iso8601(timestamp),
            'lastTradeTimestamp': undefined,
            'status': this.parseOrderStatus(this.safeString(order, 'status')),
            'symbol': symbol,
            'type': this.safeString(order, 'type'),
            'timeInForce': undefined,
            'side': this.safeStringLower(order, 'type'),
            'price': this.safeNumber(order, 'price'),
            'amount': this.safeNumber(order, 'amount'),
            'filled': this.safeNumber(order, 'filledAmount'),
            'remaining': undefined,
            'cost': undefined,
            'average': undefined,
            'trades': [],
            'fee': undefined,
        }, market);
    }
    async createOrder(symbol, type, side, amount, price = undefined, params = {}) {
        await this.loadMarkets();
        const market = this.market(symbol);
        const request = {
            'type': type,
            'srcCurrency': market['baseId'],
            'dstCurrency': market['quoteId'],
            'amount': amount,
            'side': side,
        };
        if (price !== undefined) {
            request['price'] = price;
        }
        const response = await this.request('market/orders/add', 'private', 'POST', this.extend(request, params));
        const order = this.safeDict(response, 'order', response);
        return this.parseOrder(order, market);
    }
    async cancelOrder(id, symbol = undefined, params = {}) {
        await this.loadMarkets();
        const request = {
            'order': id,
            'status': 'canceled',
        };
        const response = await this.request('market/orders/update-status', 'private', 'POST', this.extend(request, params));
        return response;
    }
    async fetchOrder(id, symbol = undefined, params = {}) {
        await this.loadMarkets();
        const orders = await this.fetchOrders(symbol, undefined, undefined, params);
        for (let i = 0; i < orders.length; i++) {
            if (orders[i]['id'] === String(id)) {
                return orders[i];
            }
        }
        throw new OrderNotFound('Order not found: ' + String(id));
    }
    async fetchOrders(symbol = undefined, since = undefined, limit = undefined, params = {}) {
        await this.loadMarkets();
        let market = undefined;
        const request = {};
        if (symbol !== undefined) {
            market = this.market(symbol);
            request['srcCurrency'] = market['baseId'];
            request['dstCurrency'] = market['quoteId'];
        }
        const response = await this.request('market/orders/list', 'private', 'POST', this.extend(request, params));
        const orders = this.safeList(response, 'orders', []);
        return this.parseOrders(orders, market, since, limit);
    }
    async fetchOpenOrders(symbol = undefined, since = undefined, limit = undefined, params = {}) {
        const request = { 'status': 'open' };
        return this.fetchOrders(symbol, since, limit, this.extend(request, params));
    }
    async fetchClosedOrders(symbol = undefined, since = undefined, limit = undefined, params = {}) {
        const request = { 'status': 'done' };
        return this.fetchOrders(symbol, since, limit, this.extend(request, params));
    }
    async fetchMyTrades(symbol = undefined, since = undefined, limit = undefined, params = {}) {
        const request = { 'status': 'done' };
        const orders = await this.fetchOrders(symbol, since, limit, this.extend(request, params));
        return orders;
    }
}
